# SparkSQL basics with SparkR  

[**Introduction to Apache Spark with R by J. A. Dianes**](https://github.com/jadianes/spark-r-notebooks)

In this notebook we will introduce basic concepts about SparkSQL with R that you can find in the [SparkR documentation](http://spark.apache.org/docs/latest/sparkr.html), applied to the [2013 American Community Survey dataset](http://www.census.gov/programs-surveys/acs/data/summary-file.html). We will do two things, read data into a SparkSQL data frame, and have a quick look at the schema and what we have read. 

## Creating a SparkSQL context

In further notebooks, we will explore our data by loading them into SparkSQL data frames. But first we need to init a SparkSQL context. The first thing we need to do is to set up some environment variables and library paths as follows. Remember to replace the value assigned to `SPARK_HOME` with your Spark home folder.  

In [1]:
# Set Spark home and R libs
Sys.setenv(SPARK_HOME='/home/cluster/spark-1.5.0-bin-hadoop2.6')
.libPaths(c(file.path(Sys.getenv('SPARK_HOME'), 'R', 'lib'), .libPaths()))

Now we can load the `SparkR` library as follows.

In [2]:
library(SparkR)


Attaching package: ‘SparkR’

The following objects are masked from ‘package:stats’:

    filter, na.omit

The following objects are masked from ‘package:base’:

    intersect, rbind, sample, subset, summary, table, transform



And now we can initialise the Spark context as [in the official documentation](http://spark.apache.org/docs/latest/sparkr.html#starting-up-sparkcontext-sqlcontext). In our case we are use a standalone Spark cluster with one master and seven workers. If you are running Spark in local node, use just `master='local'`. Additionally, we require a Spark package from Databricks to read CSV files (more on this in the next section).  

In [3]:
sc <- sparkR.init(master='spark://169.254.206.2:7077', sparkPackages="com.databricks:spark-csv_2.11:1.2.0")

Launching java with spark-submit command /home/cluster/spark-1.5.0-bin-hadoop2.6/bin/spark-submit  --packages com.databricks:spark-csv_2.11:1.2.0 sparkr-shell /tmp/RtmpyC8rUQ/backend_port3743359adb78 


And finally we can start the SparkSQL context as follows.

In [4]:
sqlContext <- sparkRSQL.init(sc)

## Creating SparkSQL data frames

### Reading CSV data using Databricks csv extension

The easiest way to get our CSV data into a [SparkSQL dataframe](http://spark.apache.org/docs/latest/sparkr.html#creating-dataframes), is by using [Databricks CSV extension](https://github.com/databricks/spark-csv) to read SparkSQL dataframes directly from csv files. In any case, remember to set the right path for your data files in the first line, ours is `/nfs/data/2013-acs/ss13husa.csv`.  

In [5]:
housing_a_file_path <- file.path('', 'nfs','data','2013-acs','ss13husa.csv')
housing_b_file_path <- file.path('', 'nfs','data','2013-acs','ss13husb.csv')

Now let's read into a SparkSQL dataframe. We need to pass four parameters in addition to the `sqlContext`:  

- The file path.  
- `header='true'` since our `csv` files have a header with the column names. 
- Indicate that we want the library to infer the schema.  
- And the source type (the Databricks package in this case).  

In [6]:
system.time(
    housing_a_df <- read.df(sqlContext, 
                        housing_a_file_path, 
                        header='true', 
                        source = "com.databricks.spark.csv", 
                        inferSchema='true')
)

   user  system elapsed 
  0.002   0.000  16.919 

Let's have a look at the inferred schema.

In [7]:
system.time(
    printSchema(housing_a_df)
)

root
 |-- RT: string (nullable = true)
 |-- SERIALNO: integer (nullable = true)
 |-- DIVISION: integer (nullable = true)
 |-- PUMA: integer (nullable = true)
 |-- REGION: integer (nullable = true)
 |-- ST: integer (nullable = true)
 |-- ADJHSG: integer (nullable = true)
 |-- ADJINC: integer (nullable = true)
 |-- WGTP: integer (nullable = true)
 |-- NP: integer (nullable = true)
 |-- TYPE: integer (nullable = true)
 |-- ACCESS: integer (nullable = true)
 |-- ACR: integer (nullable = true)
 |-- AGS: integer (nullable = true)
 |-- BATH: integer (nullable = true)
 |-- BDSP: integer (nullable = true)
 |-- BLD: integer (nullable = true)
 |-- BROADBND: integer (nullable = true)
 |-- BUS: integer (nullable = true)
 |-- COMPOTHX: integer (nullable = true)
 |-- CONP: integer (nullable = true)
 |-- DIALUP: integer (nullable = true)
 |-- DSL: integer (nullable = true)
 |-- ELEP: integer (nullable = true)
 |-- FIBEROP: integer (nullable = true)
 |-- FS: integer (nullable = true)
 |-- FULP: integer

   user  system elapsed 
  0.002   0.000   0.062 

Looks good. Let's have a look at the first few rows.  

In [8]:
head(housing_a_df)

RT SERIALNO DIVISION PUMA REGION ST  ADJHSG  ADJINC WGTP NP TYPE ACCESS ACR
1  H       84        6 2600      3  1 1000000 1007549    0  1    3     NA  NA
2  H      154        6 2500      3  1 1000000 1007549   51  4    1      1   1
3  H      156        6 1700      3  1 1000000 1007549  449  1    1      3   1
4  H      160        6 2200      3  1 1000000 1007549   16  3    1      3   3
5  H      231        6 2400      3  1 1000000 1007549   52  1    1      3  NA
6  H      286        6  900      3  1 1000000 1007549   76  1    1      1   1
  AGS BATH BDSP BLD BROADBND BUS COMPOTHX CONP DIALUP DSL ELEP FIBEROP FS FULP
1  NA   NA   NA  NA       NA  NA       NA   NA     NA  NA   NA      NA  2   NA
2  NA    1    3   2        2   2        2   NA      2   2  350       1  2    2
3  NA    1    3   2       NA   2        2   NA     NA  NA  300      NA  2    2
4   1    1    4   2       NA   2        2   NA     NA  NA  220      NA  1    2
5  NA    1    1   5       NA  NA        2   NA     NA  NA   60      NA  1    1
6  NA    1    4   2        2   2        2   NA      2   2  100       2  2    2
  GASP HANDHELD HFL INSP LAPTOP MHP MODEM MRGI MRGP MRGT MRGX OTHSVCEX REFR
1   NA       NA  NA   NA     NA  NA    NA   NA   NA   NA   NA       NA   NA
2    3        2   3  350      1  NA     2   NA   NA   NA    3        2    1
3    3        2   3  980      1  NA    NA    1  550    2    1       NA    1
4   20        1   3   NA      2  NA    NA   NA   NA   NA   NA       NA    1
5    3        2   1   NA      2  NA    NA   NA   NA   NA   NA       NA    1
6   90        2   1   50      2  NA     1    2  290    2    1        2    1
  RMSP RNTM RNTP RWAT RWATPR SATELLITE SINK SMP STOV TEL TEN TOIL VACS  VALP
1   NA   NA   NA   NA     NA        NA   NA  NA   NA  NA  NA   NA   NA    NA
2    9   NA   NA    1      9         2    1  NA    1   1   2    1   NA 25000
3    6   NA   NA    1      9        NA    1  NA    1   1   1    1   NA 80000
4    6    2  100    1      9        NA    1  NA    1   1   3    1   NA    NA
5    3    2   80    1      9        NA    1  NA    1   1   3    1   NA    NA
6    9   NA   NA    1      9         2    1  NA    1   1   1    1   NA 18000
  VEH WATP YBL FES FFINCP FGRNTP FHINCP  FINCP FPARC FSMOCP GRNTP GRPIP HHL HHT
1  NA   NA  NA  NA     NA     NA     NA     NA    NA     NA    NA    NA  NA  NA
2   3  480   2   1      1     NA      1 151000     4      0    NA    NA   1   1
3   1  700   5  NA      0     NA      1     NA    NA      1    NA    NA   1   6
4   2  360   6   8      1      0      1  11400     2     NA   370    39   1   3
5   0    1   5  NA      0      0      0     NA    NA     NA   140    43   1   6
6   1  370   2  NA      0     NA      1     NA    NA      1    NA    NA   1   6
   HINCP HUGCL HUPAC HUPAOC HUPARC KIT LNGI MULTG MV NOC NPF NPP NR NRC OCPIP
1     NA    NA    NA     NA     NA  NA   NA    NA NA  NA  NA  NA NA  NA    NA
2 151000     0     4      4      4   1    1     1  6   0   4   0  0   0     3
3  39930     0     4      4      4   1    1     1  7   0  NA   0  0   0    28
4  11400     1     2      4      2   1    1     1  4   0   3   1  0   1    NA
5   3900     0     4      4      4   1    1     1  6   0  NA   0  0   0    NA
6   5400     0     4      4      4   1    1     1  7   0  NA   0  0   0   101
  PARTNER PLM PSF R18 R60 R65 RESMODE SMOCP SMX SRNT SSMC SVAL TAXP WIF WKEXREL
1      NA  NA  NA  NA  NA  NA      NA    NA  NA   NA   NA   NA   NA  NA      NA
2       0   1   0   0   0   0       2   426  NA    0    0    1    3   2       1
3       0   1   0   0   1   0       2   926   3    0    0    1    6  NA      NA
4       0   1   0   1   1   0       2    NA  NA    0    0    0   NA   1      15
5       0   1   0   0   1   1       1    NA  NA    1    0    0   NA  NA      NA
6       0   1   0   0   1   1       1   522   3    0    0    1    3  NA      NA
  WORKSTAT FACCESSP FACRP FAGSP FBATHP FBDSP FBLDP FBROADBNDP FBUSP FCOMPOTHXP
1       NA       NA    NA    NA     NA    NA    NA         NA    NA         NA
2       

And let's count how many rows do we have in the first dataset. For that we will use `nrow` as we do with regular R data frames. Let's have a look at the documentation.  

In [9]:
?nrow

count                  package:SparkR                  R Documentation

_C_o_u_n_t

_D_e_s_c_r_i_p_t_i_o_n:

     Returns the number of rows in a DataFrame

     Aggregate function: returns the number of items in a group.

     Aggregate function: returns the number of items in a group.

_U_s_a_g_e:

     ## S4 method for signature 'DataFrame'
     count(x)
     
     ## S4 method for signature 'Column'
     count(x)
     
     ## S4 method for signature 'Column'
     n(x)
     
     n(x)
     
_A_r_g_u_m_e_n_t_s:

       x: A SparkSQL DataFrame

_S_e_e _A_l_s_o:

     Other agg_funcs: ‘approxCountDistinct’, ‘approxCountDistinct’,
     ‘approxCountDistinct’; ‘avg’, ‘avg’; ‘countDistinct’,
     ‘countDistinct’, ‘countDistinct’, ‘n_distinct’; ‘first’, ‘first’;
     ‘last’, ‘last’; ‘max’; ‘mean’; ‘min’; ‘sumDistinct’,
     ‘sumDistinct’; ‘sum’

     Other agg_funcs: ‘approxCountDistinct’, ‘approxCountDistinct’,
     ‘approxCountDistinct’; ‘avg’, ‘avg’; ‘countDistinct’,
     ‘countDistinct’, ‘countDistinct’, ‘n_distinct’; ‘first’, ‘first’;
     ‘last’, ‘last’; ‘max’; ‘mean’; ‘min’; ‘sumDistinct’,
     ‘sumDistinct’; ‘sum’

_E_x_a_m_p_l_e_s:

     ## Not run:
     
     sc <- sparkR.init()
     sqlContext <- sparkRSQL.init(sc)
     path <- "path/to/file.json"
     df <- jsonFile(sqlContext, path)
     count(df)
     ## End(Not run)
     
     ## Not run: count(df$c)

It is just a definition of `nrow` by the package *SparkR*.

In [10]:
nrow(housing_a_df)

[1] 756065

Let's read the second housing data frame and count the number of rows.

In [11]:
system.time(
    housing_b_df <- read.df(sqlContext, 
                        housing_b_file_path, 
                        header='true', 
                        source = "com.databricks.spark.csv", 
                        inferSchema='true')
)

   user  system elapsed 
  0.128   0.016   9.666 

In [12]:
print(nrow(housing_b_df))

[1] 720248


### Merging data frames

Now we can use `rbind()` as we do with regular R data frames to put both of them together. Let's have a look at the documentation.  

In [13]:
?rbind

rbind                  package:SparkR                  R Documentation

_U_n_i_o_n _t_w_o _o_r _m_o_r_e _D_a_t_a_F_r_a_m_e_s

_D_e_s_c_r_i_p_t_i_o_n:

     Returns a new DataFrame containing rows of all parameters.

_U_s_a_g_e:

     ## S4 method for signature 'DataFrame'
     rbind(x, ..., deparse.level = 1)
     
     rbind(..., deparse.level = 1)

Again, SparkR redefines many of the common R functions to work with SparkSQL data frames. Let's actually use `rbind` as follows.

In [14]:
housing_df <- rbind(housing_a_df, housing_b_df)

And let's count how many rows do we have in the complete data frame.

In [15]:
system.time(
    housing_samples <- nrow(housing_df)
)
print(housing_samples)

   user  system elapsed 
  0.001   0.000  17.206 

[1] 1476313


Finally, let's get a feeling of what is to explore data using SparkR by using the `summary` function on the data frame. Let's first have a look at the documentation.

In [16]:
?summary

describe                package:SparkR                 R Documentation

_d_e_s_c_r_i_b_e

_D_e_s_c_r_i_p_t_i_o_n:

     Computes statistics for numeric columns. If no columns are given,
     this function computes statistics for all numerical columns.

_U_s_a_g_e:

     ## S4 method for signature 'DataFrame,character'
     describe(x, col, ...)
     
     ## S4 method for signature 'DataFrame,ANY'
     describe(x)
     
     describe(x, col, ...)
     
_A_r_g_u_m_e_n_t_s:

       x: A DataFrame to be computed.

     col: A string of name

     ...: Additional expressions

_V_a_l_u_e:

     A DataFrame

_E_x_a_m_p_l_e_s:

     ## Not run:
     
     sc <- sparkR.init()
     sqlContext <- sparkRSQL.init(sc)
     path <- "path/to/file.json"
     df <- jsonFile(sqlContext, path)
     describe(df)
     describe(df, "col1")
     describe(df, "col1", "col2")
     ## End(Not run)

We can see that we are redirected to `describe`, that optionally accepts column names. Let's use it with the whole data frame, that is, the 231 columns and 1476313 rows. *Note*: We use `collect` here because the results of `describe` are given as a `DataFrame` object and we need to print them in the notebook.

In [19]:
system.time(
    housing_summary <- describe(housing_df)
)

   user  system elapsed 
  0.174   0.016 197.755 

In [20]:
collect(housing_summary)

summary      RT          SERIALNO          DIVISION              PUMA
1   count 1476313           1476313           1476313           1476313
2    mean    <NA> 746447.7617598707 5.114928202894643 4458.457590632881
3  stddev    <NA>               NaN 2.476670855019605               NaN
4     min       H                 1                 1               100
5     max       H           1492845                 9             70301
              REGION                 ST    ADJHSG    ADJINC              WGTP
1            1476313            1476313   1476313   1476313           1476313
2 2.6305945961323918   27.8048550679971 1000000.0 1007549.0 89.95933585899468
3 1.0144176501154056 15.894404191598897       NaN       NaN 80.66731786485431
4                  1                  1   1000000   1007549                 0
5                  4                 56   1000000   1007549              1829
                  NP               TYPE             ACCESS                ACR
1            1476313            1476313            1211264            1048888
2  2.122039838435345 1.1511264887594974 1.4591451574553524 1.3099472965655055
3 1.5066952326359986 0.4792639702444521 0.8182636016373733 0.5753510215594196
4                  0                  1                  1                  1
5                 20                  3                  3                  3
                 AGS                BATH               BDSP                BLD
1             240718             1328057            1328057            1328057
2 1.2790609759137248  1.0097164504234382 2.7954598334258245 2.9454202643410636
3 1.0324801853099894 0.09809200280658492 1.1626453967843118 2.0526772518004583
4                  1                   1                  0                  1
5                  6                   2                 19                 10
             BROADBND                 BUS            COMPOTHX
1              910225             1048888             1211264
2  1.5566063335988354  1.9849879110067044  1.9530713370495614
3 0.49678538927438015 0.12160068329311675 0.21148608357081813
4                   1                   1                   1
5                   2                   2                   2
                CONP              DIALUP                DSL               ELEP
1              45304              910225             910225            1211264
2 321.36614868444286  1.9596121838007086 1.6977955999890137 134.60164093046603
3 236.32344819883204 0.19686706301955062 0.4592133497896005  93.52744730809084
4                  4                   1                  1                  1
5               1600                   2                  2                600
              FIBEROP                 FS               FULP              GASP
1              910225            1359520            1211264           1211264
2   1.890910489164767  1.879595004119101 128.69548174468986 47.62741153043432
3 0.31175148670208863 0.3254345292801913   537.625361994371 71.87805326129471
4                   1                  1                  1                 1
5                   2                  2               7200               550
             HANDHELD                HFL              INSP             LAPTOP
1             1211264            1211264            847365            1211264
2  1.3795308041847194 2.2461684653386875 910.6082502817558 1.2069672672514002
3 0.48527020602919746 1.5051855327058254 915.1043685741101 0.4051318520406504
4                   1                  1                 0                  1
5                   2                  9              9100                  2
                 MHP              MODEM               MRGI               MRGP
1              55791             910225             517290             517290
2 1299.0785610582352   1.44533933917438 1.4433663902259855 1186.7885286783041
3  2208.088223691239 0.4970032315368849  0.496782280525347   859.228804208774
4                  0                  1                  1        

Or we can select individual column summaries using `select` as follows (here is [a dictionary](http://www2.census.gov/programs-surveys/acs/tech_docs/pums/data_dict/PUMSDataDict13.txt) of each column meaning) where `VALP` is the property value.  

In [25]:
collect(select(housing_summary,"VALP"))

VALP
1             859691
2 247682.84302150423
3                NaN
4                100
5            4775000

And that's it. In this notebook we have shown how to load a CSV file into an SparkSQL data frame using SparkR. We also had a look at the data loaded, mainly to the number of samples loaded, and the data summary.